File: server/adapters/anthropic.js
// Anthropic (Claude) adapter — translates the neutral conversation
// (server/conversation.js) to/from Anthropic's Messages API wire format.
// See adapters/index.js for how the three adapters fit together.

import Anthropic from '@anthropic-ai/sdk';
import { getSecret } from '../config.js';
import { systemInstructionParts } from '../prompt.js';
import { assistantTextOf } from '../conversation.js';

const MAX_TOKENS = 1024;

// What this wire format can do at all (see adapters/index.js). Images yes,
// video no — the Messages API has no video content block, which is why
// toMessages() below drops a video media item instead of translating it.
// `webSearch` is false because Anthropic's server-side web-search tool isn't
// wired up here yet; flipping it to true means adding a `searchGrounded()`
// export, and server/research.js will start using this adapter for research
// automatically the moment it does.
export const CAPABILITIES = { video: false, audio: false, vision: true, webSearch: false };

function resolveKey(entry) {
  if (entry.secretValue !== undefined) return entry.secretValue;
  if (entry.secretRef) return getSecret(entry.secretRef);
  return null;
}

function requireKey(entry) {
  const apiKey = resolveKey(entry);
  if (!apiKey) {
    const err = new Error('No API key configured.');
    err.code = 'NO_API_KEY';
    throw err;
  }
  return apiKey;
}

// `entry.baseUrl` is set only for a Custom connection whose probe (see
// server/models/probe.js) resolved to this adapter's wire shape against a
// non-Anthropic host — real Anthropic accounts never set it. Undefined
// falls through to the SDK's own default (api.anthropic.com).
function client(entry) {
  const apiKey = requireKey(entry);
  return new Anthropic({ apiKey, baseURL: entry.baseUrl || undefined });
}

// Anthropic has no video-input path in this API — a `kind: 'video'` media
// item is dropped rather than translated (see conversation.js's media field
// comment; Gemini is the one adapter with genuine native video support).
// Images translate to a proper content block either way (inline base64 or a
// plain URL source, whichever the item carries).
function mediaToBlocks(media) {
  return (media || [])
    // Images only. Everything below builds an `image` block, so anything that
    // isn't one (video, or a `document` PDF carried as inline bytes) must be
    // dropped rather than mislabelled as an image the API will reject.
    .filter((item) => !item.kind || item.kind === 'image')
    .map((item) => {
      if (item.dataBase64) {
        return { type: 'image', source: { type: 'base64', media_type: item.mimeType, data: item.dataBase64 } };
      }
      if (item.uri) {
        return { type: 'image', source: { type: 'url', url: item.uri } };
      }
      return null;
    })
    .filter(Boolean);
}

function toMessages(messages) {
  const out = [];
  for (const m of messages) {
    if (m.role === 'user' && m.media?.length) {
      out.push({
        role: 'user',
        content: [...(m.text ? [{ type: 'text', text: m.text }] : []), ...mediaToBlocks(m.media)],
      });
    } else if (m.role === 'user' && m.text) {
      out.push({ role: 'user', content: m.text });
    } else if (m.role === 'assistant') {
      // assistantTextOf(), not m.text directly — see openai-compatible.js's
      // identical comment. The raw-content shortcut below is deliberately
      // skipped when interrupted: `raw` is exactly the FULL content
      // Anthropic generated, which is precisely what must NOT be replayed
      // once a barge-in cut it short — falling through to the plain
      // spoken-text branch instead loses the raw round-trip fidelity for
      // that one turn, which is the correct trade: an interrupted turn is
      // already an edited turn, that fidelity only mattered for the
      // untruncated version.
      const spoken = assistantTextOf(m);
      if (!m.interrupted && m.raw?.adapter === 'anthropic' && m.raw.content) {
        out.push({ role: 'assistant', content: m.raw.content });
      } else if (m.toolCalls?.length) {
        out.push({
          role: 'assistant',
          content: [
            ...(spoken ? [{ type: 'text', text: spoken }] : []),
            ...m.toolCalls.map((c) => ({ type: 'tool_use', id: c.id, name: c.name, input: c.args })),
          ],
        });
      } else if (spoken) {
        out.push({ role: 'assistant', content: spoken });
      }
    } else if (m.role === 'tool' && m.toolResults?.length) {
      out.push({
        role: 'user',
        content: m.toolResults.map((r) => ({
          type: 'tool_result',
          tool_use_id: r.id,
          content: JSON.stringify(r.result),
        })),
      });
    }
  }
  return out;
}

// Our skill parameter schemas are already plain JSON Schema objects
// ({type, properties, required}), which is exactly the shape Anthropic's
// input_schema expects — no translation needed.
function toolsForAnthropic(tools) {
  return (tools || []).map((t) => ({ name: t.name, description: t.description, input_schema: t.parameters }));
}

export async function* stream(entry, messages, opts = {}) {
  const c = client(entry);
  const tools = toolsForAnthropic(opts.tools);

  // Two text blocks, not one plain string: `cache_control` on the stable
  // block marks "everything up to and including this is a cache
  // breakpoint" — a turn whose memories/pending-jobs haven't changed since
  // the last one reuses Anthropic's cached prefix (lower latency, lower
  // cost) instead of a full re-prefill on every single turn. The volatile
  // block (current time, "N minutes ago", low-confidence note) rides
  // AFTER the breakpoint with no cache_control of its own, so it never
  // needs to match byte-for-byte for the cached part to still hit — see
  // prompt.js's systemInstructionParts().
  const { stable, volatile } = systemInstructionParts(opts);
  const system = volatile
    ? [
        { type: 'text', text: stable, cache_control: { type: 'ephemeral' } },
        { type: 'text', text: volatile },
      ]
    : [{ type: 'text', text: stable, cache_control: { type: 'ephemeral' } }];

  const s = c.messages.stream(
    {
      model: entry.model,
      max_tokens: MAX_TOKENS,
      system,
      messages: toMessages(messages),
      tools,
    },
    { signal: opts.signal }
  );

  for await (const event of s) {
    if (event.type === 'content_block_delta' && event.delta?.type === 'text_delta') {
      yield { type: 'chunk', text: event.delta.text };
    }
  }

  // finalMessage() assembles the complete response (full tool_use blocks
  // included) from the events we just streamed — the documented way to get
  // both live deltas and a complete final object from one request.
  const message = await s.finalMessage();

  // message.usage was sitting on this same object the whole time and was
  // never read — see root CLAUDE.md's Cost tracking section. Emitted before
  // either terminal branch below so a tool-calling step's usage is captured
  // too, not just a turn's final answer. cache_read_input_tokens is what
  // makes the cache_control breakpoint above's actual payoff measurable for
  // the first time.
  if (message.usage) {
    yield {
      type: 'usage',
      unitKind: 'tokens',
      unitsIn: message.usage.input_tokens ?? null,
      unitsOut: message.usage.output_tokens ?? null,
      cachedIn: message.usage.cache_read_input_tokens ?? null,
      cacheWriteIn: message.usage.cache_creation_input_tokens ?? null,
      provider: 'anthropic',
      model: entry.model,
    };
  }

  const toolUses = message.content.filter((b) => b.type === 'tool_use');

  if (message.stop_reason === 'tool_use' && toolUses.length > 0) {
    yield {
      type: 'call',
      calls: toolUses.map((t) => ({ id: t.id, name: t.name, args: t.input })),
      raw: { adapter: 'anthropic', content: message.content },
    };
    return;
  }

  // Join ALL text blocks, not just the first — a reply can carry more than
  // one (e.g. text before and after a thinking/tool-adjacent block), and
  // `.find()` here used to silently drop everything after the first one:
  // the UI (which streams every text_delta above) would show the full
  // reply, but the transcript, the `done` event, and Chat History would
  // all persist only its opening fragment — invisible until a model switch
  // dropped the raw round-trip that was papering over it.
  const text = message.content
    .filter((b) => b.type === 'text')
    .map((b) => b.text)
    .join('');
  // A genuinely empty final response is a real failure, not a fake success —
  // let runner.js's existing failover machinery handle it (mark this model
  // unhealthy, try the next candidate) exactly the way a timeout already
  // does, instead of yielding a placeholder reply that gets shown/spoken as
  // if it were real and marks this model healthy. See
  // openai-compatible.js's own stream() for the same fix and its reasoning.
  if (!text) {
    const err = new Error('The model returned an empty response — try again later.');
    err.code = 'EMPTY_RESPONSE';
    throw err;
  }

  yield {
    type: 'final',
    text,
    raw: { adapter: 'anthropic', content: message.content },
  };
}

export async function testConnection(entry) {
  try {
    const c = client(entry);
    const response = await c.messages.create({
      model: entry.model,
      max_tokens: 10,
      messages: [{ role: 'user', content: 'Say "ready" and nothing else.' }],
    });
    const textBlock = response.content.find((b) => b.type === 'text');
    if (!textBlock?.text) {
      return { ok: false, error: 'The API responded but with no text — the key may be restricted.' };
    }
    return { ok: true };
  } catch (err) {
    return { ok: false, error: friendlyError(err) };
  }
}

/**
 * Lists the models this key can actually reach, via `client.models.list()`
 * (confirmed against the installed SDK's resources/models.d.ts —
 * `list(): PagePromise<ModelInfosPage, ModelInfo>` and `ModelInfo.{id,
 * display_name, max_input_tokens, capabilities}`). `PagePromise` is
 * async-iterable, so `for await` walks every page without manual cursor
 * handling. Throws on failure (bad key, network error, ...) —
 * server/models/registry.js's discoverModels() is what turns that into a
 * friendly {models, error} shape.
 */
export async function listModels(entry) {
  const c = client(entry);
  const page = await c.models.list();

  const out = [];
  for await (const m of page) {
    out.push({
      model: m.id,
      label: m.display_name || m.id,
      contextTokens: typeof m.max_input_tokens === 'number' ? m.max_input_tokens : null,
      // Anthropic's API exposes no pricing field at all — catalog.js's
      // inferBilling() fills this in from known-model heuristics instead.
      billing: null,
    });
  }
  return out;
}

export function friendlyError(err) {
  if (err?.code === 'NO_API_KEY') return 'No API key configured.';
  const nested = err?.error?.error?.message;
  if (nested) return nested;
  return err?.message || 'The key was rejected by Anthropic.';
}

File: backend/jarvis/model_system/adapters/anthropic.py
"""Anthropic — the Messages API wire format.

Two things confirmed against the real, installed SDK rather than assumed:

* **`thinking` takes a token budget**, via `{"type": "enabled",
  "budget_tokens": N}`. A `TIERS`/`VARIANT` scheme is skipped rather than
  converted into some invented number of tokens — guessing what a tier is
  "worth" here would be inventing the one thing the registry exists to stop
  being invented.
* **A raw assistant turn's own content blocks round-trip verbatim** when the
  caller supplies one (`Message.raw["content"]`) — rebuilding it from plain
  text loses information Anthropic expects back unchanged on a multi-step
  tool-calling turn.
"""

from __future__ import annotations

from typing import Any, Iterator, Mapping

from ..reasoning import ReasoningKind, ReasoningRequest
from ..request import (
    Completed, Message, ResponseFormat, StreamEvent, TextDelta, ToolCall, ToolDefinition,
    Usage,
)
from .base import AdapterError, model_id_for

NAME = "anthropic"
#: See `adapters/gemini.py`'s own flag for what this means.
SUPPORTS_REALTIME = False

MAX_TOKENS = 4096
REQUEST_TIMEOUT_S = 120.0


class NoCredential(AdapterError):
    pass


def _client(provider: Any):
    from anthropic import Anthropic

    from ..credentials import resolve

    key = resolve(provider.credential_ref)
    if not key:
        raise NoCredential("No API key configured.")
    # max_retries=0: retry policy belongs one layer up (ai/fallback.py), not
    # to each SDK — a hidden SDK-level retry would swallow exactly the
    # failure the gateway needs to see to bench a model and try the next.
    return Anthropic(api_key=key, base_url=provider.base_url or None,
                     max_retries=0, timeout=REQUEST_TIMEOUT_S)


def _media_blocks(attachments) -> list[dict[str, Any]]:
    blocks = []
    for a in attachments or ():
        if a.modality.value != "image" or not a.data:
            continue
        import base64
        blocks.append({"type": "image", "source": {
            "type": "base64", "media_type": a.mime_type or "image/png",
            "data": base64.b64encode(a.data).decode("ascii")}})
    return blocks


def to_wire(messages: tuple[Message, ...]) -> list[dict[str, Any]]:
    out: list[dict[str, Any]] = []
    for m in messages:
        if m.role == "user" and m.attachments:
            content = ([{"type": "text", "text": m.text}] if m.text else [])
            out.append({"role": "user", "content": content + _media_blocks(m.attachments)})
        elif m.role == "user" and m.text:
            out.append({"role": "user", "content": m.text})
        elif m.role == "assistant":
            raw = m.raw or {}
            if raw.get("adapter") == NAME and raw.get("content"):
                out.append({"role": "assistant", "content": raw["content"]})
            elif m.tool_calls:
                content = ([{"type": "text", "text": m.text}] if m.text else [])
                content += [{"type": "tool_use", "id": c.id, "name": c.name, "input": c.args}
                           for c in m.tool_calls]
                out.append({"role": "assistant", "content": content})
            elif m.text:
                out.append({"role": "assistant", "content": m.text})
        elif m.role == "tool" and m.tool_call_id:
            out.append({"role": "user", "content": [
                {"type": "tool_result", "tool_use_id": m.tool_call_id, "content": m.text}]})
    return out


def _system_blocks(system: str) -> list[dict[str, Any]]:
    return [{"type": "text", "text": system, "cache_control": {"type": "ephemeral"}}] if system else []


def _tools(tools: tuple[ToolDefinition, ...]) -> list[dict[str, Any]]:
    return [{"name": t.name, "description": t.description,
             "input_schema": t.parameters or {"type": "object", "properties": {}}}
            for t in tools]


def _thinking(reasoning: ReasoningRequest | None) -> dict[str, Any] | None:
    if reasoning is None or reasoning.kind is not ReasoningKind.BUDGET:
        return None
    budget = reasoning.native
    if not budget:
        return None
    return {"type": "enabled", "budget_tokens": int(budget)}


def _generation_kwargs(params: Mapping[str, Any]) -> dict[str, Any]:
    out: dict[str, Any] = {}
    if "temperature" in params:
        out["temperature"] = params["temperature"]
    if "top_p" in params:
        out["top_p"] = params["top_p"]
    if "top_k" in params:
        out["top_k"] = params["top_k"]
    if "stop_sequences" in params:
        out["stop_sequences"] = list(params["stop_sequences"])
    return out


def _usage(usage: Any) -> Usage | None:
    if usage is None:
        return None
    tokens_in = getattr(usage, "input_tokens", None)
    tokens_out = getattr(usage, "output_tokens", None)
    cached_in = getattr(usage, "cache_read_input_tokens", None)
    if tokens_in is None and tokens_out is None and cached_in is None:
        return None
    return Usage(tokens_in=tokens_in, tokens_out=tokens_out, cached_in=cached_in)


def stream(
    provider: Any,
    native_model_id: str,
    messages: tuple[Message, ...],
    system: str,
    tools: tuple[ToolDefinition, ...],
    *,
    reasoning: ReasoningRequest | None,
    params: Mapping[str, Any],
    response_format: ResponseFormat | None,
) -> Iterator[StreamEvent]:
    client = _client(provider)
    thinking = _thinking(reasoning)
    max_tokens = params.get("max_output_tokens") or MAX_TOKENS

    with client.messages.stream(
        model=model_id_for(native_model_id, reasoning),
        max_tokens=max_tokens,
        system=_system_blocks(system),
        messages=to_wire(messages),
        tools=_tools(tools),
        **_generation_kwargs(params),
        **({"thinking": thinking} if thinking else {}),
    ) as running:
        for event in running:
            if (getattr(event, "type", None) == "content_block_delta"
                    and getattr(event.delta, "type", None) == "text_delta"):
                yield TextDelta(event.delta.text)
        message = running.get_final_message()

    usage = _usage(getattr(message, "usage", None))
    blocks = [b.model_dump() if hasattr(b, "model_dump") else dict(b) for b in message.content]
    uses = [b for b in blocks if b.get("type") == "tool_use"]
    raw = {"adapter": NAME, "content": blocks}
    text = "".join(b.get("text", "") for b in blocks if b.get("type") == "text")

    if message.stop_reason == "tool_use" and uses:
        yield Completed(
            text=text,
            tool_calls=tuple(ToolCall(b["id"], b["name"], b.get("input") or {}) for b in uses),
            finish_reason="tool_calls", usage=usage, raw=raw,
        )
        return

    if not text:
        raise AdapterError("The model returned an empty response — try again later.")
    yield Completed(text=text, finish_reason="stop", usage=usage, raw=raw)


def test_connection(provider: Any, native_model_id: str) -> dict[str, Any]:
    try:
        client = _client(provider)
        message = client.messages.create(
            model=native_model_id, max_tokens=32,
            messages=[{"role": "user", "content": 'Say "ready" and nothing else.'}])
        text = "".join(getattr(b, "text", "") for b in message.content)
        if not text:
            return {"ok": False, "error": "The server responded but with no text — check the model name."}
        return {"ok": True, "error": None}
    except NoCredential:
        return {"ok": False, "error": "That connection needs an API key."}
    except Exception as err:  # noqa: BLE001
        return {"ok": False, "error": friendly_error(err)}


def discover_models(provider: Any) -> list[dict[str, Any]]:
    client = _client(provider)
    return [{"model": m.id, "label": getattr(m, "display_name", None)}
            for m in client.models.list()]


def friendly_error(err: BaseException) -> str:
    body = getattr(err, "body", None)
    if isinstance(body, dict):
        inner = body.get("error")
        if isinstance(inner, dict) and isinstance(inner.get("message"), str):
            return inner["message"]
    return str(err) or "That connection didn't work."
